Improvements to struct interop for DOTS

limoka opened the pull request:
This PR includes three improvements, which make working with DOTS based games much easier:

- Allow to inject custom structs into il2cpp runtime.
- A fix to how methods with struct pointers are generated.
- Allow Generic structs to be marked as blittable. This change could be breaking as some structs previously marked as classes now are structs. I'm performing checks if the struct is indeed blittable and making it's generic parameters constrained to `unmanaged`.

limoka committed:
Basic support of struct injection

limoka committed:
Fix methods with pointers generated wrong in interop assemblies

limoka committed:
Don't generate default constructor for structs

limoka committed:
Allow generic structs to be blittable

(cherry picked from commit f8b4eea956b50ff5cb0e86e75766c8f91b473a0f)

limoka committed:
Prevent duplicate IsUnmanagedAttribute classes from being generated

(cherry picked from commit 2ce1acbfdf4032e5e776a937eaf4e5440fe656f1)

limoka committed:
Fix formatting

limoka committed:
Cleanup

limoka committed:
Determine type usage. Add GenericBlittable struct type

limoka committed:
Generate Boxed variations for generic blittable structs

kremnev8 commented:
With the last commit, I think I fixed the generation issue Kasuromi found earlier. Now I added another kind of `TypeSpecifics`: `GenericBlittableStruct`, meaning blittability depends on used type arguments. For such structs I generate both boxed and unboxed versions, and switch them depending on the generic arguments.

limoka committed:
Append _Unboxed for blittable types instead of non blittable

js6pak commented on Il2CppInterop.Generator/Contexts/AssemblyRewriteContext.cs:
```diff
@@ -51,10 +51,12 @@ public TypeRewriteContext GetContextForNewType(TypeDefinition type)
 
     public void RegisterTypeRewrite(TypeRewriteContext context)
     {
-        if (context.OriginalType != null)
+        var exists = context.OriginalType != null && myOldTypeMap.ContainsKey(context.OriginalType);
+        if (context.OriginalType != null && !exists)
             myOldTypeMap[context.OriginalType] = context;
         myNewTypeMap[context.NewType] = context;
-        myNameTypeMap[(context.OriginalType ?? context.NewType).FullName] = context;
+        if (!exists)
+            myNameTypeMap[(context.OriginalType ?? context.NewType).FullName] = context;
     }
```
The `exists` check feels really overcomplicated, looking at the usings of it wouldn't it be easier to do something like:
```cs
public void RegisterTypeContext(TypeRewriteContext context)
{
    myNewTypeMap.Add(context.NewType, context);
}

public void RegisterTypeRewrite(TypeRewriteContext context)
{
    if (context.OriginalType != null)
        myOldTypeMap.Add(context.OriginalType, context);
    RegisterTypeContext(context);
    myNameTypeMap.Add((context.OriginalType ?? context.NewType).FullName, context);
}
```
and call `RegisterTypeContext` in `Pass12CreateGenericNonBlittableTypes`?

js6pak commented on Il2CppInterop.Generator/Passes/Pass11ComputeTypeSpecifics.cs:
```diff
@@ -29,23 +112,129 @@ private static void ComputeSpecifics(TypeRewriteContext typeContext)
+
+    internal class ParameterUsage
+    {
+        public List<List<TypeReference>> usageData;
```
Doesn't `List<TypeReference>[]` make more sense here?

js6pak commented on Il2CppInterop.Generator/Passes/Pass12CreateGenericNonBlittableTypes.cs:
```diff
@@ -0,0 +1,65 @@
+                    }
+                }
+
+                foreach (TypeRewriteContext rewriteContext in typesToAdd)
+                {
+                    assemblyContext.RegisterTypeRewrite(rewriteContext);
+                }
```
Wouldn't it be easier to iterate over `assemblyContext.OriginalTypes` above instead delaying this add?

js6pak commented on Il2CppInterop.Generator/Passes/Pass12CreateGenericNonBlittableTypes.cs:
```diff
@@ -0,0 +1,65 @@
+
+namespace Il2CppInterop.Generator.Passes
+{
+    public class Pass12CreateGenericNonBlittableTypes
```
This should be a static class in a file-scoped namespace like other passes (but I guess code style will be awful for il2cppinterop v1 no matter what)

limoka committed:
Fix incorrect generic parameter field generation

limoka committed:
Address comments

limoka committed:
Fix method generic parameter constraints

limoka committed:
Relax generic constraints which are not used in any fields

limoka committed:
Force Nullable struct to be non blittable

limoka committed:
Remove default constructor constraints

limoka committed:
Check nested types for generic parameter relaxation

limoka committed:
Fix generic constraint violations in nested types

limoka committed:
Perform generic constraint checks to ensure no type load exceptions

limoka committed:
Check for field generic parameter usage

limoka committed:
Split generic parameter code into pass

limoka committed:
Adjust generic parameter constraint logic

limoka committed:
Set Strict generic parameter constraint when found no non blittable usage

limoka committed:
Fix formatting

limoka committed:
Ensure structs are not generated as blittable only when non blittable usage is found.

limoka committed:
Generate nested types for boxed type variants

limoka committed:
Split compute specifics pass in two to prevent use cycles

limoka committed:
Prevent relaxed generic parameters from affecting blittablity.

limoka committed:
Revert forcing Nullable struct to be non blittable

limoka committed:
Fix invalid constructor being generated in unboxed struct variants

limoka committed:
Generate boolean fields in generic structs as a custom struct.

limoka committed:
Allow fields to be ignored

ds5678 commented:
Hello! I am open to reviewing this pull request if you rebase it.

limoka commented:
I can do that sometime later, however recently I'm rather busy with Core Keeper's 1.0 release.
On top of that I no longer have a test game. Core Keeper no longer uses il2cpp, and the only other game I know of that uses this PR is VRising.

limoka commented:
Also if I had to guess there is a lot of changes that would be needed because of #124 for to work properly.

ds5678 commented:
> Also if I had to guess there is a lot of changes that would be needed because of #124 for this to work properly.

That is correct. The switch to AsmResolver required many changes. In some ways, it was a rewrite of the generator.

limoka commented:
Do you have any Cecil to AsmResolver migration guides perhaps? I'm not familiar with the latter.

ds5678 commented:
They are very similar, though I don't have any guides.

limoka committed:
Port generic struct generation changes to use AsmResolver

limoka committed:
Fix generation inconsistencies with previous version

limoka committed:
fix formatting issues

limoka commented:
@ds5678 Hello, I have rebased and updated the PR in order to work with AsmResolver as you asked.

I have checked that the output of this version of PR is consistent or very close to original PR output. To verify this I used your tool mentioned in your #124 PR.
For these checks I have used VRising game, which is one of other known to me il2cpp + DOTS games (Core Keeper has switched to mono, so I'm unable to test on it)
On VRising's ProjectM.dll all 7k types match, only ~3k methods differ from ~55k, only ~3k fields differ from ~54k. Of the differences I found, most are either fixed issues or irrelevant differences. 

Additionally I have installed a few mods from VRising Thunderstore (Bloodstone, VCF). From my quick test, all tested mods are working.

limoka committed:
tidy up code

ds5678 commented on Il2CppInterop.Runtime/Attributes/HideFromIl2CppAttribute.cs:
```diff
@@ -6,7 +6,7 @@ namespace Il2CppInterop.Runtime.Attributes;
 ///     This attribute indicates that the target should not be exposed to IL2CPP in injected classes
 /// </summary>
 [AttributeUsage(AttributeTargets.Method | AttributeTargets.Constructor | AttributeTargets.Property |
-                AttributeTargets.Event)]
```
Not injecting all the instance fields of a struct sounds like a bad idea.

limoka replied:
Well, you are definitely correct that this is a bad idea.

The reason I did this was that you cannot inject a generic type.  And I needed to have my job structs contain a handle type. Like this:
```cs
public struct ExampleJob
{
    //These fields can be used as is
    public float inputData;
    public NativeArray<Entity> entities;
    public EntityCommandBuffer ecb;

    // I can't inject this type, nor can I use original type (Because it doesn't have code compiled for CustomComponentCD)
    // So I hide the field
    [HideFromIl2Cpp] 
    public ModComponentDataFromEntity<CustomComponentCD> inputFromEntity;
}
```

In order not to have crazy issues I had to hack Job Scheduler code, access internal structs and manually adjust the struct size for the code to work.
Or simply it's a giant hack, but it did work.

As such I see two solutions:
1. Remove this ability and implement generic injection or some way to use generic types with types not used by the game code (This one would probably be a huge thing, and would simplify a lot of things I had to do)
2. Keep this for now, but perhaps warn if encountered to inform the modder about consequences of doing this

ds5678 commented on Il2CppInterop.Runtime/Injection/ClassInjector.cs:
```diff
@@ -198,8 +205,8 @@ public static void RegisterTypeInIl2Cpp(Type type, RegisterTypeOptions options)
         // Initialize the vtable of all base types (Class::Init is recursive internally)
         InjectorHelpers.ClassInit(baseClassPointer.ClassPointer);
 
-        if (baseClassPointer.ValueType || baseClassPointer.EnumType)
-            throw new ArgumentException($"Base class {baseType} is value type and can't be inherited from");
```
Does `baseClassPointer.ValueType` return `true` for `System.ValueType`? If not, this shouldn't be changed.

limoka replied:
I'm pretty sure it does return `true`. This check being removed is exactly why structs can be injected. This is big part of this PR.

Specifically structs are needed because Unity's ECS (Entity Component System) heavily relies on structs to hold game state. By injecting custom ones we can add our own.

ds5678 commented on Il2CppInterop.Generator/Utils/GenericParameterContext.cs:
```diff
@@ -0,0 +1,45 @@
+
+namespace Il2CppInterop.Generator.Utils;
+
+public struct GenericParameterContext
```
What's wrong with `AsmResolver.DotNet.Signatures.GenericContext`?

limoka replied:
The problem is that It handles a different problem. Specifically `GenericContext` allows to resolve a type from a `GenericParamterSignature`. 
But what I need here is to resolve `GenericParameter` (Which has all the information about the parameter including constraints and attributes) from `GenericParamterSignature` (Which is a weird type that doesn't even know who it belongs to).

This is all caused by the weird design of AsmResolver. I have read some messages on their discord describing why `GenericParamterSignature` is as it is. However the fact metadata stores it this way, doesn't really justify complicating handling of `GenericParamterSignature` by making them glorified ints. I found Cecil's approach of `GenericParameters` having all context at all times much more convenient, and the code was written around that. 

So `GenericParameterContext` was needed to implement the logic with AsmResolver. Perhaps maybe Washi can take a look at this code and say how would they do it.

ds5678 commented on Il2CppInterop.Generator/Runners/InteropAssemblyGenerator.cs:
```diff
@@ -67,29 +67,39 @@ public void Run(GeneratorOptions options)
+            Pass11ComputeGenericParameterSpecifics.DoPass(rewriteContext);
+        }
+
         using (new TimingCookie("Computing struct blittability"))
         {
-            Pass11ComputeTypeSpecifics.DoPass(rewriteContext);
+            Pass12ComputeTypeSpecifics.DoPass(rewriteContext);
+        }
+
+        using (new TimingCookie("Creating unboxed struct types"))
+        {
+            Pass13CreateGenericNonBlittableTypes.DoPass(rewriteContext);
         }
 
         using (new TimingCookie("Filling typedefs"))
         {
-            Pass12FillTypedefs.DoPass(rewriteContext);
+            Pass14FillTypedefs.DoPass(rewriteContext);
         }
 
         using (new TimingCookie("Filling generic constraints"))
         {
-            Pass13FillGenericConstraints.DoPass(rewriteContext);
+            Pass15FillGenericConstraints.DoPass(rewriteContext);
         }
 
         using (new TimingCookie("Creating members"))
         {
-            Pass15GenerateMemberContexts.DoPass(rewriteContext);
+            Pass16GenerateMemberContexts.DoPass(rewriteContext);
         }
 
         using (new TimingCookie("Scanning method cross-references"))
         {
-            Pass16ScanMethodRefs.DoPass(rewriteContext, options);
+            Pass17ScanMethodRefs.DoPass(rewriteContext, options);
         }
```
Please make this part of pass 11 instead of renaming the other passes. It fits the theme of `ComputeTypeSpecifics`.

limoka replied:
Sure. I just felt that `ComputeTypeSpecifics` step code size was growing too big

limoka committed:
Move Compute Generic Parameter Specifics to compute type specifics pass

limoka committed:
Change pass numbers to be closer to original numbering

limoka committed:
Format code
